pkg/ipmanager/ipam/controller.go

```diff
@@ -0,0 +1,312 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+func NewController(spec *SubmarinerIpamControllerSpecification, config *InformerConfigStruct) (*Controller, error) {
+	exclusionMap := make(map[string]bool)
+	for _, v := range spec.ExcludeNS {
+		exclusionMap[v] = true
+	}
+	pool, err := NewIpPool(spec.GlobalCIDR)
+	if err != nil {
+		return nil, err
+	}
+
+	ipamController := &Controller{
+		kubeClientSet:    config.KubeClientSet,
+		serviceWorkqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Services"),
+		servicesSynced:   config.ServiceInformer.Informer().HasSynced,
+		podWorkqueue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Pods"),
+		podsSynced:       config.PodInformer.Informer().HasSynced,
+
+		excludeNamespaces: exclusionMap,
+		pool: pool,
+	}
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: ipamController.enqueueService,
+		UpdateFunc: func(old, new interface{}) {
+			ipamController.enqueueService(new)
+		},
+		DeleteFunc: ipamController.handleRemovedService,
+	}, handlerResync)
+	config.PodInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: ipamController.enqueuePod,
+		UpdateFunc: func(old, new interface{}) {
+			ipamController.handleUpdatePod(old, new)
+		},
+		DeleteFunc: ipamController.handleRemovedPod,
+	}, handlerResync)
+
+	return ipamController, nil
+}
+
+func (i *Controller) Run(stopCh <-chan struct{}) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting IPAM Controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, i.servicesSynced, i.podsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Info("Starting workers")
+	go wait.Until(i.runServiceWorker, time.Second, stopCh)
+	go wait.Until(i.runPodWorker, time.Second, stopCh)
+	wg.Wait()
+	<-stopCh
+	klog.Info("Shutting down workers")
+
+	return nil
+}
+
+func (i *Controller) runServiceWorker() {
+	for i.processNextService() {
+	}
+}
+
+func (i *Controller) runPodWorker() {
+	for i.processNextPod() {
+	}
+}
+
+func (i *Controller) processNextService() bool {
+	obj, shutdown := i.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer i.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			i.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+		if i.excludeNamespaces[ns] {
+			i.serviceWorkqueue.Forget(obj)
+			return nil
+		}
```

Comment:
Before enqueing the event in the WorkQueue, we are already checking it against the excluded namespaces. So, this looks redundant.

Reply:
Taken care in refactoring.